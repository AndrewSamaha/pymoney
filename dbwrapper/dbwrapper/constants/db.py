dbfilepath = '../db/money.sqlite'
rawCsvPath = '../db/rawTransactionHistory'
connectionString = f"sqlite:///{dbfilepath}"

def getConnectionString(env='local'):
    connectionString = {
        'local': f"sqlite:///../db/money.sqlite",
        'test': 'sqlite:///tests/mock_data/mockdb.sqlite'
    }
    if env.lower() in connectionString:
        return connectionString[env.lower()]
    return connectionString['test']

def getSeedFolder(env='local'):
    seedFolder={
        'local':    '../db/seed_data/',
        'test':     './tests/mock_data/seed_data/',
        'default':  './testsmock_data/seed_data/'
    }
    if env.lower() in seedFolder:
        return seedFolder[env.lower()]
    return seedFolder['default']