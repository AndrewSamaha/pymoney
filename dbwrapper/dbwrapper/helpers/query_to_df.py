import pandas as pd
#from dbwrapper.helpers.readCsv import setDateColumns
from . import setDateColumns
#from dbwrapper.helpers.readCsv import setDateColumns

def query_to_df(query, convert_date_columns=True):
    result = query.all()
    column_names = [ x['name'] for x in query.column_descriptions]
    df = pd.DataFrame(result, columns=column_names)
    if convert_date_columns:
        df = setDateColumns(df)
    return df

