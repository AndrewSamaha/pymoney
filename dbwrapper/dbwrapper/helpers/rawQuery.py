def rawQuery(sql, engine):
    Session = sessionmaker(bind=engine)
    session = Session()
    result = session.execute(text(sql))

    # Process the result
    for idx, row in enumerate(result):
        print(row)
        #if idx > 5: break

    # Close the result set and the connection
    session.close()

def rawQueryDf(sql, engine):
    Session = sessionmaker(bind=engine)
    session = Session()
    df = pd.read_sql(sql=text(sql), con=session.connection())   
    session.close()
    return df
