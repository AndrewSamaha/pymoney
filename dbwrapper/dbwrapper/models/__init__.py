# dbwrapper.models.__init__.py

from .Account import Account
from .Base import Base
from .Load import Load
from .Transaction import Transaction, TransactionFinal
