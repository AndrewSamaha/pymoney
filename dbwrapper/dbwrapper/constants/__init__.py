# data-pipeline.dbwrapper.constants.__init__.py

from .db import connectionString, getConnectionString, getSeedFolder, rawCsvPath, dbfilepath
