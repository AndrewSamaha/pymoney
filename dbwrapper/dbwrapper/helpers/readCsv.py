import pandas as pd
import os
import re
from datetime import datetime
from sqlalchemy import text

from ..constants import rawCsvPath
from ..models import Load
from . import setDateColumns, getDateColumns, hash, hashType

def openOneCsv(path):
    date_columns = getDateColumns(path)
    df = pd.read_csv(path, index_col=False, parse_dates=date_columns)
    if "Balance" in df.columns:
        df["Balance"] = pd.to_numeric(df["Balance"], errors="coerce")
    df[hashType] = df.apply(hash, axis=1)
    df['stagingLevel'] = 'allTransactions'
    
    return df

def prepCheckingDf(df, loadID):
    df = df.rename(columns={
        'Details': 'details',
        'Description': 'description',
        'Amount': 'amount',
        'Type': 'type',
        'Balance': 'balance',
        'Check or Slip #': 'checkOrSlipNumber',
        'Post Date': 'postDate',
        'Transaction Date': 'transactionDate',
        'Category': 'category'
    })
    df = df.drop(['seenTime'], errors='ignore')
    df['loadID'] = loadID
    return df

def getAccountHistory(path, accountId, session, rawCsvPath=rawCsvPath):
    df = None
    accountPath = f"{rawCsvPath}/{path}"
    for file in os.listdir(accountPath):
        fileAndPath = f"{rawCsvPath}/{path}/{file}"
        filenameParts = file.split('.')
        filename = filenameParts[:len(filenameParts)-1][0]

        newDf = openOneCsv(fileAndPath)

        if 'Posting Date' in newDf.columns:
            newDf['Post Date'] = newDf['Posting Date']
            newDf = newDf.drop(axis=1, columns=['Posting Date'])
        
        if session is not None:
            matchingLoads = pd.read_sql(sql=text(f"select * from loads WHERE fullFilePath = '{fileAndPath}'"), con=session.connection()) 

            if matchingLoads.shape[0] > 0:
                print(f"  skipping {matchingLoads.shape[0]}")
                continue
            
            load = Load(accountId=accountId, fullFilePath=fileAndPath)
            session.add(load)
            session.commit()
            newDf['accountId'] = accountId
            preppedDf = prepCheckingDf(newDf, load.id)

        if df is None:
            df = preppedDf
        else:
            df = pd.concat([df, preppedDf])

    if df is None:
        return None

    # Convert Date Strings to Date
    dateColumns = ['Post Date','Transaction Date']
    for dateColumn in dateColumns:
        if dateColumn in df.columns:
            df[dateColumn] = pd.to_datetime(df[dateColumn], format='%Y-%m-%d') # %m-%d-%Y

    # sort by Post Date and then sha
    df = df.sort_values(by=['postDate', 'sha256'], ascending=False)
    
    return df
