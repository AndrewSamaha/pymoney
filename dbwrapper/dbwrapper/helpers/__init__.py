# data-pipeline.dbwrapper.helpers.__init__.py

from .dates import getDateColumns, setDateColumns
from .hash import hash, hashLength, hashType
from .query_to_df import query_to_df
from .rawQuery import rawQuery, rawQueryDf
from .sqlEngine import engine
from .summarize import summarize
from .seed import seed_worker, getSeedEngine
from .readCsv import getAccountHistory
