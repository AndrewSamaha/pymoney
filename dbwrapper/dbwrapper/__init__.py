from .constants import connectionString

from .models.Base import Base
from .models.Account import Account
from .models.Transaction import TransactionFinal, Transaction
from .models.Load import Load

from .helpers.sqlEngine import engine
from .helpers.query_to_df import query_to_df

from .helpers.readCsv import setDateColumns
from .helpers import rawQuery, rawQueryDf

from .crud.read.select_summary_df import select_summary_df

