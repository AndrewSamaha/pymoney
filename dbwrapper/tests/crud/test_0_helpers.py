import pytest
import os
from dbwrapper import testvalue
from dbwrapper.constants import getConnectionString, getSeedFolder
from dbwrapper.helpers import seed_worker, getSeedEngine
from dbwrapper.models import Base

envs = [
    'Test',
    'local',
    'default'
]

def test_getConnectionString(snapshot):
    for env in envs:
        cString = getConnectionString(env)
        envCStringPair = f"{env} == {cString}"
        assert envCStringPair == snapshot

def test_getSeedFolder(snapshot):
    for env in envs:
        result = getSeedFolder(env)
        inputResultPair = f"{env} == {result}"
        assert inputResultPair == snapshot
