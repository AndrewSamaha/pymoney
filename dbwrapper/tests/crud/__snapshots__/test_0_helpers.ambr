# serializer version: 1
# name: test_getConnectionString
  'Test == sqlite:///tests/mock_data/mockdb.sqlite'
# ---
# name: test_getConnectionString.1
  'local == sqlite:///../db/money.sqlite'
# ---
# name: test_getConnectionString.2
  'default == sqlite:///tests/mock_data/mockdb.sqlite'
# ---
# name: test_getSeedFolder
  'Test == ./tests/mock_data/seed_data/'
# ---
# name: test_getSeedFolder.1
  'local == ../db/seed_data/'
# ---
# name: test_getSeedFolder.2
  'default == ./testsmock_data/seed_data/'
# ---
